### ISSUE TITLE
MAX - Weapon raycasts causing FPS drops and CPU spikes (Experimental/DX12 build)

### ISSUE THREAD

**Issue by fearlesswee** [TheGameCreators/GameGuruRepo#3523]
I've noticed stutters and frame drops when firing weapons, most notably the shotgun which fires a scatter of 6 or so raycasts at once. On an empty testing map this creates a somewhat-noticeable but admittedly brief/minor stutter, however on a populated map this stutter is extremely jarring and makes the already "meh" combat/shooting feel even more janky and awful. With the shotgun it's most egregious because it practically "locks up" the shotgun and you can see each raycast/pellet hitting one-at-a-time, almost like a burst-fire weapon rather than a scatter of pellets all-at-once. (This does seem to "hang up" the weapon system as the fire rate of the shotgun is severely lowered as it struggles to slowly cough out each raycast one after the other before it can continue functioning to ready the next shot.)

SYSTEM SPECS:
CPU: Ryzen 3600 6-core processor
GPU: NVIDIA GeForce GTX 1660 SUPER
Storage: 512GB SSD
RAM: 8 GB (plus 6 GB virtual memory)

MAX performance is "fine" otherwise beyond some (lack of) optimization issues.

**Comment by davetgc** [comment]
Can you please post a video of what is happening to you. I couldn't replicate this.

**Comment by fearlesswee** [comment]
[https://streamable.com/ps1uc6](https://streamable.com/ps1uc6)

It can be difficult to see the "hitches" and slower rate-of-fire across the video, but in gameplay it's VERY noticeable and very jarring. I fire a few shots into the air as I've noticed the hitching/stuttering doesn't happen when the weapons are fired into the air or otherwise hit something that doesn't create a bullethole, like terrain/entities that don't have a material properly set (half of the stock assets lol)

**Comment by fearlesswee** [comment]
I'll experiment with giving the shotgun something like 30+ pellets in the gunspec so that the effect is more pronounced and comes across much more plainly in the recording.

**Comment by fearlesswee** [comment]
Further testing has revealed a **wealth** of bugs with the weapon system as a whole:

1. Shotgun spread patterns are _very_ strange; rather than fire them in a completely random scatter as one would logically expect, they fire in a fixed "ring" pattern with a single pellet hitting dead-center every time. This means it's more effective to purposefully aim off-center from a target as it will mean more pellets will actually hit the target, as if you aim for center-mass you will only hit that single center pellet and do pitiful damage.
2. If you fire a weapon and then unequip it quickly, it will leave a lingering muzzle flash light that remains until the weapon is re-equipped and fired again.
3. If you fire a weapon and then unequip it quickly, any "pending" bullet holes will wait until you re-equip the weapon to appear, which is extremely strange.

[https://streamable.com/emtewg](https://streamable.com/emtewg)

**Comment by fearlesswee** [comment]
In my experimentation I've also discovered there seems to be a hard-coded limit on weapon spread, so you can't make it go beyond a certain amount (which is quite low ). Why is this restriction here? Shouldn't we, as the developer, have more control over our own games' weapons?

**Comment by fearlesswee** [comment]
Updated the title as after some testing on a different system, I've found that firing a weapon raycast causes severe and noticeable spikes in CPU times which drops FPS by up to _HALF_, which is a HUGE issue.

I created an "empty" biome and dropped the camera distance to be just barely enough to see the gun, to ensure any CPU drops aren't due to any rendering and just CPU-bound stuff. The inbuilt performance metrics report "MAX - General" as taking about 2.25-3 ms when idle, while firing the weapon into the sky increases this time to around 11-15 ms, which drops frames by about 2-3 FPS and causes some annoying but not major microstuttering. When hitting a placed zombie character, it increases to **55-60 ms** and can drop up to **30 FPS.** 

In an FPS-oriented game maker, having the very act of shooting a gun demolishing performance is not a good thing.

**Comment by fearlesswee** [comment]
Further testing has revealed the more severe stuttering and FPS drops are only in the experimental build. Rolling back to the public build has caused the weapons to revert to only causing minor stutter instead of entirely freezing the software briefly and slicing frames in half. I've also found projectiles (the stock grenades) cause constant slowdowns (down to a measly 12 FPS!) while they are active and rolling around, with performance only picking back up once they detonate. This leads me to believe something in the DX12 build related to raycasts and projectile physics (if they use raycasts in any way) is broken/unoptimized and causing these massive slowdowns. This should be sorted out with the utmost priority imho.

**Comment by LeeBamberTGC** [comment]
@fearlesswee DX12 and new Wicked merge taking longer than I anticipated. When you see the new and shiny build on EXP out the door, let me know how it fairs with your tests, thanks!

**Comment by fearlesswee** [comment]
Can confirm I still get these severe performance hits as of the latest experimental EXP build. (GameGuru MAX Build 2022.10.14 )

**Comment by plemsoft** [comment]
@fearlesswee Is this still a issue in the latest build ? , i remember that Lee optimized the raycasting some time back :)

**Comment by fearlesswee** [comment]
> Is this still a issue in the latest build ? , i remember that Lee optimized the raycasting some time back :)

Went ahead and updated to latest EXP build, plopped down 15 or so zombies on an empty level and an assault rifle with a bunch of ammo, and did similar tests to before; shooting in the sky, shooting the ground, and then finally engaging with the zombies. The stuttering/hitching seems to be entirely gone when shooting in the air or ground, but when shooting zombies the first hit caused a serious hitch; the game briefly froze up for like 0.1 seconds, but after that first shot it seems to run absolutely fine with only a few tiny hiccups.

The "first shot hitching" seems be somewhat consistent; if I boot up MAX, go into a test game, and fire a shot at an enemy, that first shot will almost always cause a serious CPU usage spike and freezing, but then it's mostly fine after that. The "first shot" only seems to "reset" once MAX is fully closed; if I go into test game again on the same level, or even load another level and test that, the hitching doesn't happen on that first shot. Can't really say why this might be the case, but certainly hope my finding helps. Even then, it's not a major issue since it's only that first shot and the hitch is so brief.

In short; it's seriously improved a ton, but there's still one minor annoyance left.

**Comment by plemsoft** [comment]
@fearlesswee Strange i dont see that here , perhaps its the terrain streaming when you first start the level , this will drop the fps while streaming for a few second , after that it will settle at a steady fps , and on my first shot there was no fps drops ? could you try to see if its the same for you , move  the player start marker away from the zombies , let the fps settle and try to fire your first shot , do you still see this fps drop ?. if so could you try to open the profiler and when making the first shoot take a screenshot, that would help , Thanks :)

**Comment by fearlesswee** [comment]
@plemsoft 

[https://streamable.com/mvl4jx](https://streamable.com/mvl4jx)

You can see the spike in CPU usage in the profiler, along with a brief hitch or freeze in the software. In video it doesn't look _too_ bad, but when actually playing the game it feels quite awful. When recording this I realized it actually happens on the first _few_ shots, and not just the single first shot. I also fire into the air and ground first to show it only happens when shooting enemies. Could it be that it's loading the blood particle textures into memory, and that's causing it to "halt" and wait until the blood is loaded? 

In the video I also had waited in the level 60 seconds (1 minute) before starting recording to ensure there was no terrain streaming shenanigans going on (which, was also the case before when I was first doing these tests lol)

Hope this helps!

**Comment by plemsoft** [comment]
@fearlesswee You found it , yes it was the decal loading and creating of mipmaps, this has been fixed and will be in the next update, Thanks for your help :)

**Comment by fearlesswee** [comment]
Awesome! Nice to see I'm not totally clueless lol! Happy to have helped, and even happier this has been fixed! I'll be sure to test it right away in the next exp update to confirm 👍

**Comment by fearlesswee** [comment]
Going through my old issues and seeing what is/isn't fixed. Checking on this issue the issue seems almost entirely resolved, but it would appear there are still a few small edge cases where firing a gun will cause a brief but noticeable stutter/jitter. From what I can tell it happens mostly when shooting at characters, especially when there are multiple of them close together.

**Comment by LeeBamberTGC** [comment]
@fearlesswee There will be some performance boosts along with the bug fixing work, so if you find a nice small stock asset way to demonstrate a performance drop/spike, attach here so I can see on my slow-end rig and see what is taking up the cycles.  I agree all events in the game should stay smooth no matter what action is going on.

**Comment by fearlesswee** [comment]
Doing some tests on the latest experimental build, beyond the usual jitters/performance problems MAX has in general, the jitters and hitching caused by firing weapons themselves _seems_ to be addressed. Firing a shotgun at the aztec walls was my usual test, which revealed the previously present issues are no longer present. Also did a runthrough of the canyon offensive demo level (which revealed a couple of unrelated issues I'm going to report) and didn't notice any weapon-caused stutters or hitching. I think this can be safely marked as "fixed" for now, but I'll reopen it if the issue re-reveals itself.
